src/Extsload.sol

```diff
@@ -7,46 +7,52 @@ import {IExtsload} from "./interfaces/IExtsload.sol";
+    function extsload(bytes32 slot) external view returns (bytes32) {
         /// @solidity memory-safe-assembly
         assembly {
-            value := sload(slot)
+            mstore(0, sload(slot))
+            return(0, 0x20)
         }
     }
 
     /// @inheritdoc IExtsload
     function extsload(bytes32 startSlot, uint256 nSlots) external view returns (bytes memory) {
-        bytes memory value = new bytes(32 * nSlots);
-
         /// @solidity memory-safe-assembly
         assembly {
-            for { let i := 0 } lt(i, nSlots) { i := add(i, 1) } {
-                mstore(add(value, mul(add(i, 1), 32)), sload(add(startSlot, i)))
+            // The abi offset of dynamic array in the returndata is 32.
+            mstore(0, 0x20)
+            // A left bit-shift of 5 is equivalent to multiplying by 32 but costs less gas.
+            mstore(0x20, shl(5, nSlots))
+            let end := add(0x40, shl(5, nSlots))
+            for { let memptr := 0x40 } 1 {} {
+                mstore(memptr, sload(startSlot))
+                memptr := add(memptr, 0x20)
+                startSlot := add(startSlot, 1)
+                if iszero(lt(memptr, end)) { break }
             }
+            // The end offset is also the length of the returndata.
+            return(0, end)
         }
-
-        return value;
     }
 
     /// @inheritdoc IExtsload
     function extsload(bytes32[] calldata slots) external view returns (bytes32[] memory) {
         // since the function is external and enters a new call context and exits right
         // after execution, Solidity's memory management convention can be disregarded
         // and a direct slice of memory can be returned
-        assembly ("memory-safe") {
-            // abi offset for dynamic array
-            mstore(0, 0x20)
-            mstore(0x20, slots.length)
-            let end := add(0x40, shl(5, slots.length))
-            let memptr := 0x40
-            let calldataptr := slots.offset
-            for {} 1 {} {
-                mstore(memptr, sload(calldataload(calldataptr)))
+        /// @solidity memory-safe-assembly
+        assembly {
+            // Copy the abi offset of dynamic array and the length of the array to the memory.
+            calldatacopy(0, 0x04, 0x40)
+            // Return values will start at 64 while calldata offset is 68.
+            for { let memptr := 0x40 } 1 {} {
+                mstore(memptr, sload(calldataload(slots.offset)))
                 memptr := add(memptr, 0x20)
-                calldataptr := add(calldataptr, 0x20)
-                if iszero(lt(memptr, end)) { break }
+                slots.offset := add(slots.offset, 0x20)
+                if iszero(lt(slots.offset, calldatasize())) { break }
             }
-            return(0, end)
+            // The size of the returndata is the size of calldata minus 4 for the selector.
+            return(0, sub(calldatasize(), 0x04))
```

Comment:
i dont think thats always true? you can send any amount of calldata you want to a function and it just parses the bytes that it needs

Reply:
i know youre checking this - just adding so i dont forget. This would also mean you can no longer use the same break condition. I think you need to use the actual slots.length to iterate through without relying on calldata size

Reply:
You're right. Reverted.